File: src/tauri/keychain.ts
// Thin wrapper around the Rust keychain commands (PLAN.md #4). Credentials
// never touch SQLite or a plaintext file — only the OS keychain, via these
// three Tauri invokes.

import { invoke } from "@tauri-apps/api/core";

import type { Credentials } from "../lib/types";

export async function keychainSet(
  connectionId: string,
  credentials: Credentials,
): Promise<void> {
  await invoke("keychain_set", {
    connectionId,
    accessKey: credentials.accessKey,
    secretKey: credentials.secretKey,
  });
}

export async function keychainGet(connectionId: string): Promise<Credentials | null> {
  return invoke<Credentials | null>("keychain_get", { connectionId });
}

export async function keychainDelete(connectionId: string): Promise<void> {
  await invoke("keychain_delete", { connectionId });
}

File: src-tauri/src/keychain.rs
use lopload_native::keychain::{self, Credentials};

#[tauri::command]
pub fn keychain_set(
    connection_id: String,
    access_key: String,
    secret_key: String,
) -> Result<(), String> {
    keychain::set(
        &connection_id,
        &Credentials {
            access_key,
            secret_key,
        },
    )
}

#[tauri::command]
pub fn keychain_get(connection_id: String) -> Result<Credentials, String> {
    keychain::get(&connection_id)
}

#[tauri::command]
pub fn keychain_delete(connection_id: String) -> Result<(), String> {
    keychain::delete(&connection_id)
}
